nototools/gsub_diff.py

```diff
@@ -44,41 +44,144 @@ def __init__(self, file_a, file_b, output_lines=20):
-            contents = re.findall(contents_rx % (name, name), text, re.S)
-            assert len(contents) == 1, 'Multiple %s features in %s' % (
-                name, filename)
-            contents = contents[0]
-            for lhs, rhs in re.findall(rule_rx, contents):
-                rules.add((name, lhs, rhs))
-        return rules
+            features.add(name)
+        return list(features)
+
+    def _get_feature_content(self, text, feature):
+        contents_rx = r'feature %s {(.*?)} %s;'
+        contents = re.findall(contents_rx % (feature, feature), text, re.S)
+        return contents
+
+    def _get_lookups_rules(self, text, content, feature):
+        """Ignore rules which use "'". These are contextual and not in
+        lookups 1-4"""
+        rule_rx = r"[^C] sub (.*[^\']) (by|from) (.*);"
+        rules = re.findall(rule_rx, content)
+        parsed_rules = self._parse_gsub_rules(rules, feature)
+        return parsed_rules
+
+    def _parse_gsub_rules(self, rules, feature):
+        """
+        Parse GSUB sub LookupTypes 1, 2, 3, 4, 7. Return list of tuples with
+        the following tuple sequence.
+
+        (feature, [input glyphs], operator, [output glyphs])
+
+        Type 1 Single Sub:
+        sub a by a.sc;
+        sub b by b.sc;
+        [
+            (feat, ['a'], 'by' ['a.sc']),
+            (feat, ['b'], 'by' ['b.cs'])
+        ]
+
+
+        Type 2 Multiple Sub:
+        sub f_f by f f;
+        sub f_f_i by f f i;
+        [
+            (feat, ['f_f'], 'by', ['f', 'f']),
+            (feat, ['f_f_i'], 'by', ['f', 'f', 'i'])
+        ]
+
+        Type 3 Alternative Sub:
+        sub ampersand from [ampersand.1 ampersand.2 ampersand.3];
+            [
+                (feat, ['ampersand'], 'from', ['ampersand.1']),
+                (feat, ['ampersand'], 'from', ['ampersand.2']),
+                (feat, ['ampersand'], 'from', ['ampersand.3'])
+            ]
+
+        Type 4 Ligature Sub:
+        sub f f by f_f;
+        sub f f i by f_f_i;
+        [
+            (feat, ['f', 'f'] 'by' ['f_f]),
+            (feat, ['f', 'f', 'i'] 'by' ['f_f_i'])
+        ]
+
+        http://www.adobe.com/devnet/opentype/afdko/topic_feature_file_syntax.html#4.e
+        """
+        parsed = []
+        for idx, (left, op, right) in enumerate(rules):
+
+            left_group, right_group = [], []
+            if left.startswith('[') and left.endswith(']'):
+                left = self._gsub_rule_group_to_string(left)
+
+            if right.startswith('[') and right.endswith(']'):
+                right = self._gsub_rule_group_to_string(right)
+
+            if op == 'by': # parse LookupType 1, 2, 4
+                parsed.append((feature, left.split(), op, right.split()))
+            elif op == 'from': # parse LookupType 3
+                for glyph in right.split(): # 'a.alt a.sc' -> ['a.alt', 'a.sc']
+                    parsed.append((feature, left.split(), op, [glyph]))
```

Comment:
`left.split()` shouldn't be necessary since in lookuptype 3 alternate substitutions the format is `substitute <glyph> from <glyphclass>;` so the left slot can contain only one glyph. A `left.strip()` would do.